tst/com/amazon/corretto/crypto/provider/test/EvpSignatureSpecificTest.java

```diff
@@ -364,21 +369,173 @@ public void testRsaWithoutCrtParams() throws Exception {
+        assertThrows(IllegalStateException.class, () -> signer.setParameter(spec2));
+        assertThrows(IllegalStateException.class, () -> verifier.setParameter(spec2));
+        assertTrue(verifier.verify(signer.sign()));
+
+        // Finally, assert that we can set params after sign/verify reinit
+        signer.setParameter(spec2);
+        verifier.setParameter(spec2);
+        assertPssParamsEqual(spec2, signer.getParameters().getParameterSpec(PSSParameterSpec.class));
+        assertPssParamsEqual(spec2, verifier.getParameters().getParameterSpec(PSSParameterSpec.class));
+    }
+
+    private static void assertPssParamsEqual(PSSParameterSpec s1, PSSParameterSpec s2) {
+        assertEquals(s1.getDigestAlgorithm(), s2.getDigestAlgorithm());
+        assertEquals(s1.getDigestAlgorithm(), s2.getDigestAlgorithm());
+        assertEquals(
+            ((MGF1ParameterSpec) s1.getMGFParameters()).getDigestAlgorithm(),
+            ((MGF1ParameterSpec) s2.getMGFParameters()).getDigestAlgorithm()
+        );
+    }
+
+    @Test
+    void testRsaPSSInitSmallKeyAfterSetParameterLargeSaltThrows() throws Exception {
+        final int minimallySecureKeyLen = 2048;
+        final int smallKeySize = 2048 / 4;
+        final Signature signer = Signature.getInstance("RSASSA-PSS", NATIVE_PROVIDER);
+        final int largeSaltLen = 2 * (smallKeySize / 8);
+        final PSSParameterSpec spec = new PSSParameterSpec("SHA-1", "MGF1", MGF1ParameterSpec.SHA1, largeSaltLen, 1);
+
+        // Set PSS params _before_ initializing for signer. No access to key length yet, so assuming a default of 2048.
+        signer.setParameter(spec);
+
+        // Now, initialize the signature with a smaller key
+        final KeyPairGenerator kg = KeyPairGenerator.getInstance("RSA", NATIVE_PROVIDER);
+        kg.initialize(smallKeySize);
+        KeyPair pair = kg.generateKeyPair();
+        signer.initSign(pair.getPrivate());
+
+        // Assert that doFinal (i.e. the native method called in signer.sign()) detects overly large salt len and throws
+        byte[] shortMessage = new byte[] { (byte) 0xDE, (byte) 0xAD, (byte) 0xBE, (byte) 0xEF };
+        signer.update(shortMessage);
+        assertThrows(SignatureException.class, () -> signer.sign());
+
+        // After re-initializing with minimally secure RSA key, longer salt len set earlier should be fine.
+        kg.initialize(minimallySecureKeyLen);
+        pair = kg.generateKeyPair();
+        signer.initSign(pair.getPrivate());
+        signer.update(shortMessage);
+        final byte[] signature = signer.sign();
+
+        // For completeness, also verify the data and signature. We don't test verification with the smaller key
+        // size because generating a siganture with oversized salt length + undersized key is not possible. If no
+        // signature can be generated, there's nothing nothing to verify.
+        final Signature verifier = Signature.getInstance("RSASSA-PSS", NATIVE_PROVIDER);
```

Comment:
What happens when you *try* to validate the signature though? Yeah, it cannot actually exist. So generate a valid signature (with reasonable salt) and then configure your validation logic to use it invalid salt. Do you get a `false` (because it isn't valid), a `SignatureException` (correctly detecting there is something wrong with the configuration/setup) or a `CryptoRuntimeException` because it fails in an unexpected way?

Reply:
Slightly different case than what you describe above, but same result as [mentioned in our prior thread](https://github.com/corretto/amazon-corretto-crypto-provider/pull/202#discussion_r898268843). It simply returns false. To be totally explicit, with this diff applied to the PR's [current HEAD commit](https://github.com/corretto/amazon-corretto-crypto-provider/commit/ed7ec940e9d8a56ba8075a966930e1757c2109ea):

```
diff --git a/tst/com/amazon/corretto/crypto/provider/test/EvpSignatureSpecificTest.java b/tst/com/amazon/corretto/crypto/provider/test/EvpSignatureSpecificTest.java
index 3c8c612..82d4e13 100644
--- a/tst/com/amazon/corretto/crypto/provider/test/EvpSignatureSpecificTest.java
+++ b/tst/com/amazon/corretto/crypto/provider/test/EvpSignatureSpecificTest.java
@@ -518,6 +518,23 @@ public final class EvpSignatureSpecificTest {
         verifier.initVerify(pair.getPublic());
         verifier.update(shortMessage);
         assertTrue(verifier.verify(signature));
+
+        // Generate a valid signature with an undersized key but minimal salt length. Then, bump the salt length
+        // for the verifier instance and try to verify the signature. The verification attempt will return false
+        // instead of throwing. Apparently AWS-LC doesn't perform relevant validation upon verification.
+        PSSParameterSpec spec2 = new PSSParameterSpec("SHA-1", "MGF1", MGF1ParameterSpec.SHA1, /*saltLen*/20, 1);
+        kg.initialize(smallKeySize);
+        pair = kg.generateKeyPair();
+        signer.setParameter(spec2);
+        signer.initSign(pair.getPrivate());
+        signer.update(shortMessage);
+        final byte[] signature2 = signer.sign();
+
+        spec2 = new PSSParameterSpec("SHA-1", "MGF1", MGF1ParameterSpec.SHA1, largeSaltLen, 1);
+        verifier.setParameter(spec2);
+        verifier.initVerify(pair.getPublic());
+        verifier.update(shortMessage);
+        assertFalse(verifier.verify(signature));
     }
 
 

```

All tests pass:

```
root@79bf3ce255b6:/accp# ./gradlew cmake_clean && ./gradlew single_test -DSINGLE_TEST=com.amazon.corretto.crypto.provider.test.EvpSignatureSpecificTest
...
[         0 tests failed          ]

[100%] Built target check-junit-single

BUILD SUCCESSFUL in 2m 25s
```